### PR TITLE
Add publisher for wheel joints.

### PR DESCRIPTION
**Pull request by shokoofeh**
- In the dead_rockening node, wheel joint states are computed from encoder data and published on `joint_states` topic. 
- fixes https://github.com/g/grizzly/issues/2

### CODE REVIEW

File: grizzly_motion/include/grizzly_motion/dead_reckoning.h [hunk 1]
```diff
@@ -80,6 +81,12 @@ class DeadReckoning
   geometry_msgs::Point position_; 
   geometry_msgs::Twist twist_; 
   double yaw_;
+  bool initialize;
+  std::vector<std::string> joint_name_;
+  std::vector<double> joint_pos_;
+  std::vector<double> joint_vel_;
+  std::vector<>double joint_eff_;
```
Comment (mikepurvis) [inline grizzly_motion/include/grizzly_motion/dead_reckoning.h]: Whoops, this isn't going to compile, I don't think.

Also, are these variables necessary? What about just a single instance of JointState which you populate? It contains vectors for the variable-length fields; you just resize them.

File: grizzly_motion/src/dead_reckoning.cpp [hunk 2]
```diff
@@ -67,14 +90,25 @@ bool DeadReckoning::next(const grizzly_msgs::DriveConstPtr& encoders, nav_msgs::
     Eigen::Map<Eigen::MatrixXd> poseCov(odom->pose.covariance.data(), 6, 6);
     Eigen::Map<Eigen::MatrixXd> twistCov(odom->twist.covariance.data(), 6, 6);
 
-    if(twist_.linear.x <= 1e-3 && twist_.angular.z <= 1e-3) {
+    if(fabs(twist_.linear.x) <= 1e-3 && fabs(twist_.angular.z) <= 1e-3) {
```
Comment (mikepurvis) [inline grizzly_motion/src/dead_reckoning.cpp:90]: Ah, thank you for this fix.

File: grizzly_motion/src/dead_reckoning.cpp [hunk 3]
```diff
@@ -34,10 +34,27 @@ using Eigen::Vector2f;
  * commands. Currently very naive, but in the future may provide some further
  * intelligence, though not closed-loop control.
  */
-bool DeadReckoning::next(const grizzly_msgs::DriveConstPtr& encoders, nav_msgs::Odometry* odom)
+void wrap_rad(double& r){ r = fabs(r) < M_PI ? r : (r -= copysignf(1.0, r) * 2.0 * M_PI);}
```
Comment (mikepurvis) [inline grizzly_motion/src/dead_reckoning.cpp]: Looks sketchy. Please remove this and use `angles::normalize_angle` instead: http://docs.ros.org/api/angles/html/

(And watch out about inserting functions between another function and its comment block...)

File: grizzly_motion/src/dead_reckoning.cpp [hunk 4]
```diff
@@ -55,7 +72,13 @@ bool DeadReckoning::next(const grizzly_msgs::DriveConstPtr& encoders, nav_msgs::
     // Update linear and angular velocity
     twist_.linear.x = vels.mean();
     twist_.angular.z = (vels[1] - vels[0]) / width_;
-    
+
+    //update joint's velocity
+    joint_vel_[0] = encoders->rear_left;
+    joint_vel_[1] = encoders->rear_right;
+    joint_vel_[2] = encoders->front_left;
+    joint_vel_[3] = encoders->front_right;
```
Comment (mikepurvis) [inline grizzly_motion/src/dead_reckoning.cpp]: I should have alerted you to this earlier, but you can use [vectorFromDriveMessage](http://docs.ros.org/hydro/api/grizzly_msgs/html/namespacegrizzly__msgs.html#aeb1f2b9e670fb638bc9e2accb4738873) here to eliminate this logic.

**Comment by shokoofeh** [comment]
Thanks @mikepurvis, I made your proposed changes. 

Now joint names are read from Drive msg. so they'll be `front_left`, `front_right`, `rear_left` and `rear_right` as oppose to `joint_back_left_wheel` so on.

**Comment by mikepurvis** [comment]
Ack, those aren't going to match the URDF now, though. The [joint names are](https://github.com/g/grizzly/blob/9c3a503bfe5b0843c57ab7c38b721277a2a44486/grizzly_description/urdf/wheel.urdf.xacro#L124), eg: `joint_front_left_wheel`.

Unfortunately, the URDF uses front/back instead of front/rear. But I'd consider that a bug in the URDF. Consider updating the URDF (and gazebo plugin) as well: https://github.com/g/grizzly/blob/9c3a503bfe5b0843c57ab7c38b721277a2a44486/grizzly_description/urdf/base.urdf.xacro#L183

@paulbovbel Can you confirm that you've gone to front/rear naming convention for Husky?

**Comment by paulbovbel** [comment]
Yes, mirroring jackal config.

**Comment by mikepurvis** [comment]
LGTM

**Pull request merged**